### PR TITLE
Registration/login implementation

### PR DESCRIPTION
**Pull request by Justas293**
Using in-memory authorization for now.

### CODE REVIEW

File: server/ShoppingSystem/src/main/java/lt/mif/ise/domain/Role.java [hunk 1]
```diff
@@ -0,0 +1,39 @@
+package lt.mif.ise.domain;
+
+import javax.persistence.*;
+import java.util.Set;
+
+@Entity
+@Table(name = "role")
+public class Role {
+    private Long id;
```
Comment (sarunassp) [inline server/ShoppingSystem/src/main/java/lt/mif/ise/domain/Role.java]: We used string for id in product, should keep that consistent

File: server/ShoppingSystem/src/main/java/lt/mif/ise/domain/User.java [hunk 2]
```diff
@@ -0,0 +1,60 @@
+package lt.mif.ise.domain;
+
+import javax.persistence.*;
+import java.util.Set;
+
+@Entity
+@Table(name="user")
+public class User {
+    private Long id;
```
Comment (sarunassp) [inline server/ShoppingSystem/src/main/java/lt/mif/ise/domain/User.java]: string id

File: server/ShoppingSystem/src/main/java/lt/mif/ise/domain/Role.java [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+    private String name;
+    private Set<User> users;
```
Comment (sarunassp) [inline server/ShoppingSystem/src/main/java/lt/mif/ise/domain/Role.java]: user points to roles, roles points to user?

File: server/ShoppingSystem/src/main/java/lt/mif/ise/service/impl/UserServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
+import org.springframework.stereotype.Service;
+
+import java.util.HashSet;
+
+@Service
+public class UserServiceImpl implements UserService {
+    @Autowired
+    private UserRepository userRepository;
```
Comment (sarunassp) [inline server/ShoppingSystem/src/main/java/lt/mif/ise/service/impl/UserServiceImpl.java:19]: should probably use final for these, but i dont know if thats possible with autowire. Its a good practice so the variables cannot be changed after setting them

File: server/ShoppingSystem/src/main/java/lt/mif/ise/service/impl/UserDetailsServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+    @Transactional(readOnly = true)
+    public UserDetails loadUserByUsername(String username) throws UsernameNotFoundException {
+        User user = userRepository.findByUsername(username);
+
+        Set<GrantedAuthority> grantedAuthorities = new HashSet<>();
+        for (Role role : user.getRoles()){
+            grantedAuthorities.add(new SimpleGrantedAuthority(role.getName()));
+        }
+
+        return new org.springframework.security.core.userdetails.User(user.getUsername(), user.getPassword(), grantedAuthorities);
```
Comment (sarunassp) [inline server/ShoppingSystem/src/main/java/lt/mif/ise/service/impl/UserDetailsServiceImpl.java:33]: check if java has import alias, this looks ugly

File: server/ShoppingSystem/src/main/java/lt/mif/ise/service/impl/UserDetailsServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+public class UserDetailsServiceImpl implements UserDetailsService{
+    @Autowired
+    private UserRepository userRepository;
+
+    @Override
+    @Transactional(readOnly = true)
+    public UserDetails loadUserByUsername(String username) throws UsernameNotFoundException {
+        User user = userRepository.findByUsername(username);
+
+        Set<GrantedAuthority> grantedAuthorities = new HashSet<>();
```
Comment (sarunassp) [inline server/ShoppingSystem/src/main/java/lt/mif/ise/service/impl/UserDetailsServiceImpl.java:28]: creating a hashset and casting it to a set?

File: server/ShoppingSystem/src/main/java/lt/mif/ise/validator/UserValidator.java [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+    public boolean supports(Class<?> aClass) {
+        return User.class.equals(aClass);
+    }
+
+    @Override
+    public void validate(Object o, Errors errors) {
+        User user = (User) o;
+
+        ValidationUtils.rejectIfEmptyOrWhitespace(errors, "username", "NotEmpty");
+        if (user.getUsername().length() < 6 || user.getUsername().length() > 32) {
```
Comment (sarunassp) [inline server/ShoppingSystem/src/main/java/lt/mif/ise/validator/UserValidator.java]: should probably move these numbers to a config

File: server/ShoppingSystem/src/main/java/lt/mif/ise/validator/UserValidator.java [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+    @Autowired
+    private UserService userService;
+
+    @Override
+    public boolean supports(Class<?> aClass) {
+        return User.class.equals(aClass);
+    }
+
+    @Override
+    public void validate(Object o, Errors errors) {
```
Comment (sarunassp) [inline server/ShoppingSystem/src/main/java/lt/mif/ise/validator/UserValidator.java]: maybe there's a validator which uses generics?

File: server/ShoppingSystem/src/main/java/lt/mif/ise/domain/Role.java [hunk 9]
```diff
@@ -0,0 +1,39 @@
+package lt.mif.ise.domain;
+
+import javax.persistence.*;
+import java.util.Set;
+
+@Entity
+@Table(name = "role")
```
Comment (sarunassp) [inline server/ShoppingSystem/src/main/java/lt/mif/ise/domain/Role.java:10]: do we need to write table names ourselves? didnt do that with product

File: server/ShoppingSystem/src/main/java/lt/mif/ise/domain/User.java [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+    private String username;
+    private String password;
+    private String passwordConfirm;
```
Comment (sarunassp) [inline server/ShoppingSystem/src/main/java/lt/mif/ise/domain/User.java]: shouldnt keep password confirmation thingy in database

File: server/ShoppingSystem/src/main/java/lt/mif/ise/rest/controller/UserController.java [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+    public String registration(){
+        return "registration";
+    }
+
+    @RequestMapping(value="/login", method = RequestMethod.GET)
+    public String login(){
+        return "login";
+    }
+
+    @RequestMapping(value={"/", "/welcome"}, method = RequestMethod.GET)
```
Comment (sarunassp) [inline server/ShoppingSystem/src/main/java/lt/mif/ise/rest/controller/UserController.java]: _{"/", "/welcome"}_ wat

File: server/ShoppingSystem/src/main/java/lt/mif/ise/rest/controller/UserController.java [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+import lt.mif.ise.domain.User;
+import lt.mif.ise.service.SecurityService;
+import lt.mif.ise.service.UserService;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RequestMethod;
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+public class UserController {
```
Comment (sarunassp) [inline server/ShoppingSystem/src/main/java/lt/mif/ise/rest/controller/UserController.java]: doesnt have base path

File: server/ShoppingSystem/src/main/java/lt/mif/ise/service/impl/UserServiceImpl.java [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+    @Autowired
+    private RoleRepository roleRepository;
+    @Autowired
+    private BCryptPasswordEncoder bCryptPasswordEncoder;
+
+    @Override
+    public void save(User user){
+        user.setPassword(bCryptPasswordEncoder.encode(user.getPassword()));
+        user.setRoles(new HashSet<>(roleRepository.findAll()));
```
Comment (sarunassp) [inline server/ShoppingSystem/src/main/java/lt/mif/ise/service/impl/UserServiceImpl.java]: setroles takes a set why cast?

File: server/ShoppingSystem/src/main/java/lt/mif/ise/validator/UserValidator.java [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+        User user = (User) o;
+
+        ValidationUtils.rejectIfEmptyOrWhitespace(errors, "username", "NotEmpty");
+        if (user.getUsername().length() < 6 || user.getUsername().length() > 32) {
+            errors.rejectValue("username", "Size.userForm.username");
+        }
+        if (userService.findByUsername(user.getUsername()) != null) {
```
Comment (sarunassp) [inline server/ShoppingSystem/src/main/java/lt/mif/ise/validator/UserValidator.java]: does it actually return null if it doesnt find the user?

File: server/ShoppingSystem/src/main/java/lt/mif/ise/validator/UserValidator.java [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+    @Override
+    public boolean supports(Class<?> aClass) {
+        return User.class.equals(aClass);
+    }
+
+    @Override
+    public void validate(Object o, Errors errors) {
+        User user = (User) o;
+
+        ValidationUtils.rejectIfEmptyOrWhitespace(errors, "username", "NotEmpty");
```
Comment (sarunassp) [inline server/ShoppingSystem/src/main/java/lt/mif/ise/validator/UserValidator.java]: wat? youre checking if a contstant string is empty

File: server/ShoppingSystem/src/main/java/lt/mif/ise/service/impl/UserServiceImpl.java (hunk 4 above)
Comment (arvyy) [reply server/ShoppingSystem/src/main/java/lt/mif/ise/service/impl/UserServiceImpl.java:19]: No final, and even `private` is often omitted for clarity.

File: server/ShoppingSystem/src/main/java/lt/mif/ise/service/impl/UserDetailsServiceImpl.java (hunk 6 above)
Comment (arvyy) [reply server/ShoppingSystem/src/main/java/lt/mif/ise/service/impl/UserDetailsServiceImpl.java:28]: Set is interface, hashset is class implementation. All ok

File: server/ShoppingSystem/src/main/java/lt/mif/ise/service/impl/UserServiceImpl.java (hunk 4 above)
Comment (Justas293) [reply server/ShoppingSystem/src/main/java/lt/mif/ise/service/impl/UserServiceImpl.java:19]: final and autowired are contradictory.

File: server/ShoppingSystem/src/main/java/lt/mif/ise/domain/Role.java [hunk 16]
```diff
@@ -6,17 +6,18 @@
 @Entity
 @Table(name = "role")
 public class Role {
-    private Long id;
+
+    private String id;
+
```
Comment (sarunassp) [inline server/ShoppingSystem/src/main/java/lt/mif/ise/domain/Role.java]: remove new line

Comment (sarunassp) [inline server/ShoppingSystem/src/main/java/lt/mif/ise/domain/Role.java]: remove new line

File: server/ShoppingSystem/src/main/java/lt/mif/ise/jpa/RoleRepository.java [hunk 17]
```diff
@@ -3,6 +3,8 @@
 import lt.mif.ise.domain.Role;
 import org.springframework.data.jpa.repository.JpaRepository;
 
-public interface RoleRepository extends JpaRepository<Role, Long> {
+import java.util.Set;
```
Comment (sarunassp) [inline server/ShoppingSystem/src/main/java/lt/mif/ise/jpa/RoleRepository.java]: you really like randon new lines :D

File: server/ShoppingSystem/src/main/java/lt/mif/ise/rest/controller/UserController.java [hunk 18]
```diff
@@ -1,37 +1,42 @@
...
-import lt.mif.ise.service.SecurityService;
 import lt.mif.ise.service.UserService;
 import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.security.access.prepost.PreAuthorize;
+import org.springframework.web.bind.annotation.RequestBody;
 import org.springframework.web.bind.annotation.RequestMapping;
 import org.springframework.web.bind.annotation.RequestMethod;
 import org.springframework.web.bind.annotation.RestController;
 
+@RequestMapping("/users")
```
Comment (sarunassp) [inline server/ShoppingSystem/src/main/java/lt/mif/ise/rest/controller/UserController.java]: api/users

File: server/ShoppingSystem/src/main/java/lt/mif/ise/rest/controller/UserController.java [hunk 19]
```diff
@@ -1,37 +1,42 @@
...
+    @PreAuthorize(("hasAnyRole('USER')"))
+    @RequestMapping(value="/test/secured/user", method = RequestMethod.GET)
+    public String testUserAuthorization(){
+        return "welcome USER";
     }
 
-    @RequestMapping(value={"/", "/welcome"}, method = RequestMethod.GET)
-    public String welcome(){
-        return "welcome";
+    @RequestMapping(value = "/test/user", method = RequestMethod.GET)
```
Comment (sarunassp) [inline server/ShoppingSystem/src/main/java/lt/mif/ise/rest/controller/UserController.java]: its probably better to have / at the beginning not here but on base path

File: server/ShoppingSystem/src/main/java/lt/mif/ise/security/WebSecurityConfig.java [hunk 20]
```diff
@@ -24,26 +30,21 @@ public BCryptPasswordEncoder bCryptPasswordEncoder(){
...
-                .and()
-                    .formLogin()
-                        //.loginPage("/login")
-                            .permitAll()
-                .and()
-                    .logout()
-                .permitAll();
+        http.csrf().disable();
+        http.authorizeRequests()
+                .antMatchers("**/secured/**").authenticated()
```
Comment (sarunassp) [inline server/ShoppingSystem/src/main/java/lt/mif/ise/security/WebSecurityConfig.java]: nitpicking but use 4 spaces for tab here

File: server/ShoppingSystem/src/main/java/lt/mif/ise/security/WebSecurityConfig.java [hunk 21]
```diff
@@ -24,26 +30,21 @@ public BCryptPasswordEncoder bCryptPasswordEncoder(){
...
     }
 
     @Autowired
     public void configureGlobal(AuthenticationManagerBuilder auth) throws Exception {
-        auth.userDetailsService(userDetailsService).passwordEncoder(bCryptPasswordEncoder());
+        //auth.userDetailsService(userDetailsService).passwordEncoder(bCryptPasswordEncoder());
+
+        //temporary in memory authentification, since we dont have normal database yet
+        auth.inMemoryAuthentication()
+                .withUser("admin").password(bCryptPasswordEncoder().encode("admin")).roles("ADMIN")
```
Comment (sarunassp) [inline server/ShoppingSystem/src/main/java/lt/mif/ise/security/WebSecurityConfig.java]: 4 spaces

**Review by sarunassp — CHANGES_REQUESTED** [review]

**Pull request merged**